### PR TITLE
feat: add exposeInIsolatedWorld(worldId, key, api) to contextBridge

### PR DESCRIPTION
**Pull request by akshaydeo**
#### Description of Change

This PR adds a new function, exposeInIsolatedWorld to contextBridge. With this change, custom APIs can be exposed to Isolated worlds without any hacks when `contextIsolation` is enabled.

Fixes https://github.com/electron/electron/issues/32588

cc @miniak 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added `contextBridge.exposeInIsolatedWorld(worldId, key, api)` to expose an API to an `isolatedWorld` within a renderer from a preload script.

### CODE REVIEW

File: docs/api/context-bridge.md [hunk 1]
```diff
@@ -84,6 +90,47 @@ contextBridge.exposeInMainWorld(
 )
 ```
 
+An example of complex API exposed in an isolated world is shown below:
```
Comment (MarshallOfSound) [inline docs/api/context-bridge.md:93]: I don't think we need a _complex_ example here, an incredibly simple example would suffice.  E.g. only expose a single method in the isolated world

File: lib/renderer/api/context-bridge.ts [hunk 2]
```diff
@@ -1,13 +1,21 @@
 const binding = process._linkedBinding('electron_renderer_context_bridge');
 
 const checkContextIsolationEnabled = () => {
-  if (!process.contextIsolated) throw new Error('contextBridge API can only be used when contextIsolation is enabled');
+  if (!process.contextIsolated) {
```
Comment (MarshallOfSound) [inline lib/renderer/api/context-bridge.ts:4]: It looks like you've run a code formatter on our files that _isn't_ our own linting rules, can you undo these changes to make this diff easier to review.

**Review by MarshallOfSound — CHANGES_REQUESTED** [review]
As per my comment above, please undo all automated formatting changes in this PR not caused by our own linter as it makes the diff almost impossible to review accurately if a lot of the changed lines don't actually contain changes

File: docs/api/context-bridge.md (hunk 1 above)
Comment (akshaydeo) [reply docs/api/context-bridge.md]: Added a simple example for `exposeInIsolatedWorld`

**Comment by akshaydeo** [comment]
I've updated this PR to remove all lining differences

File: spec-main/api-context-bridge-spec.ts [hunk 3]
```diff
@@ -62,977 +62,2051 @@ describe('contextBridge', () => {
...
+          const result = await callWithBindings(async (root: any) => {
+            const fn = () => null;
+            return root.example(fn) === fn;
+          });
+          expect(result).equal(true);
+        });
+
+        it('should properly handle errors thrown in proxied functions', async () => {
+          await makeBindingWindow(() => {
+            contextBridge.exposeInMainWorld('example', () => {
```
Comment (RaisinTen) [inline spec-main/api-context-bridge-spec.ts:370]: the exposeInMainWorld tests still look as if these have been formatted. IIUC, there shouldn't be a diff for the existing tests.

File: spec-main/api-context-bridge-spec.ts [hunk 4]
```diff
@@ -1227,7 +2301,9 @@ describe('ContextBridgeMutability', () => {
     const appProcess = cp.spawn(process.execPath, ['--enable-logging', '--enable-features=ContextBridgeMutability', appPath]);
 
     let output = '';
-    appProcess.stdout.on('data', data => { output += data; });
+    appProcess.stdout.on('data', data => {
```
Comment (RaisinTen) [inline spec-main/api-context-bridge-spec.ts:2304]: this one and the one below too seem to be formatted differently

**Review by RaisinTen — COMMENTED** [review]
I think some parts of `spec-main/api-context-bridge-spec.ts` still has the formatting changes.

File: spec-main/api-context-bridge-spec.ts (hunk 3 above)
Comment (akshaydeo) [reply spec-main/api-context-bridge-spec.ts:370]: I have contextualized a few initializations and added a grouping over existing test cases; that's why there might also be some differences in the current test cases.
I will cross-check the styling once again 🙇‍♂️

**Comment by akshaydeo** [comment]
The diff in the existing content from spec file is because I had to create a separate `describe` blob to segregate `exposeInMainWorld` and `exposeInIsolatedWorld` scenarios 🙇‍♂️

File: docs/api/context-bridge.md [hunk 5]
```diff
@@ -46,11 +46,17 @@ The `contextBridge` module has the following methods:
 * `apiKey` string - The key to inject the API onto `window` with.  The API will be accessible on `window[apiKey]`.
 * `api` any - Your API, more information on what this API can be and how it works is available below.
 
+### `contextBridge.exposeInIsolatedWorld(worldId, apiKey, api)`
+
+* `worldId` Integer - The ID of the world to inject the API into. This has to be an existing world.
```
Comment (nornagon) [inline docs/api/context-bridge.md:51]: iirc `999` is the world ID that the preload uses. What happens if you try to expose a method in world `999`? Should we do anything to prevent that? Should we at least note something about that?

**Comment by nornagon** [comment]
Haven't reviewed the code but API LGTM.

File: docs/api/context-bridge.md (hunk 5 above)
Comment (akshaydeo) [reply docs/api/context-bridge.md:51]: Exposing functionality in preload context could be a reasonable expectation, but it's not the concern for the functionality of exposeInIsolatedWorld.

I will mention using 999 would expose the object for preload context in the documentation :man-bowing 🙇‍♂️

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 6]
```diff
@@ -607,6 +607,59 @@ void ExposeAPIInMainWorld(v8::Isolate* isolate,
   }
 }
 
+void ExposeAPIInIsolatedWorld(v8::Isolate* isolate,
+                              const int world_id,
+                              const std::string& key,
+                              v8::Local<v8::Value> api,
+                              gin_helper::Arguments* args) {
+  TRACE_EVENT1("electron", "ContextBridge::ExposeAPIInIsolatedWorld", "worldId",
+               world_id);
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:616]: For consistency with `ExposeAPIInMainWorld`, let's continue to use the `key` argument.
```suggestion
  TRACE_EVENT2("electron", "ContextBridge::ExposeAPIInIsolatedWorld", "key", key, "worldId",
               world_id);
```

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 7]
```diff
@@ -607,6 +607,59 @@ void ExposeAPIInMainWorld(v8::Isolate* isolate,
...
+                                  isolated_context->Global());
+
+  if (isolated.Has(key)) {
+    args->ThrowError(
+        "Cannot bind an API on top of an existing property on the window "
+        "object");
+    return;
+  }
+
+  v8::Local<v8::Context> isolated_context2 = frame->GetScriptContextFromWorldId(
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:637]: Maybe rename to `electron_isolated_context`

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 8]
```diff
@@ -717,6 +770,8 @@ void Initialize(v8::Local<v8::Object> exports,
   v8::Isolate* isolate = context->GetIsolate();
   gin_helper::Dictionary dict(isolate, exports);
   dict.SetMethod("exposeAPIInMainWorld", &electron::api::ExposeAPIInMainWorld);
+  dict.SetMethod("exposeAPIInIsolatedWorld",
+                 &electron::api::ExposeAPIInIsolatedWorld);
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:774]: Since both `exposeAPIInMainWorld` and `exposeAPIInIsolatedWorld` have almost the same implementation, I'd be in favor of exporting a single `exposeAPIInWorld` method.

The internal JS could then become
```ts
const contextBridge: Electron.ContextBridge = {
  exposeInMainWorld: (key: string, api: any) => {
    checkContextIsolationEnabled();
    return binding.exposeAPIInWorld(0, key, api);
  },
  exposeInIsolatedWorld: (worldId: number, key: string, api: any) => {
    checkContextIsolationEnabled();
    return binding.exposeAPIInWorld(worldId, key, api);
  }
};
```

File: spec-main/api-context-bridge-spec.ts [hunk 9]
```diff
@@ -60,1165 +60,2149 @@ describe('contextBridge', () => {
...
+    };
 
-      it('should proxy promises and reject with the correct value', async () => {
-        await makeBindingWindow(() => {
-          contextBridge.exposeInMainWorld('example', Promise.reject(new Error('i-rejected')));
+    generateTests(true);
+    generateTests(false);
+  });
+
+  describe('exposeInIsolatedWorld', () => {
```
Comment (samuelmaddock) [inline spec-main/api-context-bridge-spec.ts:1225]: Given the similarity in implementation, I'd be in favor of only adding a simple test for `exposeInIsolatedWorld` rather than duplicating all of those for `exposeInMainWorld`.

This would cut down on testing time and redundancy.

**Review by samuelmaddock — CHANGES_REQUESTED** [review]
API LGTM

Had some feedback about the code.

File: docs/api/context-bridge.md (hunk 5 above)
Comment (samuelmaddock) [reply docs/api/context-bridge.md]: Please also include a test which calls `exposeInIsolatedWorld` with that world ID.

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 10]
```diff
@@ -560,50 +560,57 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
   CHECK(render_frame);
   auto* frame = render_frame->GetWebFrame();
   CHECK(frame);
+
   v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
-  gin_helper::Dictionary global(main_context->GetIsolate(),
-                                main_context->Global());
+  v8::Local<v8::Context> context =
+      world_id == 0 ? main_context
+                    : frame->GetScriptContextFromWorldId(isolate, world_id);
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:579]:
```suggestion
  v8::Local<v8::Context> target_context =
      world_id == 0 ? frame->MainWorldScriptContext()
                    : frame->GetScriptContextFromWorldId(isolate, world_id);
```

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 11]
```diff
@@ -560,50 +560,57 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
 
-  v8::Local<v8::Context> isolated_context = frame->GetScriptContextFromWorldId(
-      args->isolate(), WorldIDs::ISOLATED_WORLD_ID);
+  v8::Local<v8::Context> electron_isolated_context =
+      frame->GetScriptContextFromWorldId(args->isolate(),
+                                         WorldIDs::ISOLATED_WORLD_ID);
 
   {
     context_bridge::ObjectCache object_cache;
     v8::Context::Scope main_context_scope(main_context);
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:596]: The `v8::Context::Scope` needs be constructed using the context which will be modified. In my other comment, I used `target_context` to refer to it.

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 12]
```diff
@@ -560,50 +560,57 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
   CHECK(frame);
+
   v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
-  gin_helper::Dictionary global(main_context->GetIsolate(),
-                                main_context->Global());
+  v8::Local<v8::Context> context =
+      world_id == 0 ? main_context
+                    : frame->GetScriptContextFromWorldId(isolate, world_id);
+
+  gin_helper::Dictionary properties(context->GetIsolate(), context->Global());
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:581]: nitpick: I think `global` still makes sense to use for naming here.

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 13]
```diff
@@ -560,50 +560,57 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
 
     if (base::FeatureList::IsEnabled(features::kContextBridgeMutability)) {
-      global.Set(key, proxy);
+      isolated.Set(key, proxy);
       return;
     }
 
     if (proxy->IsObject() && !proxy->IsTypedArray() &&
-        !DeepFreeze(proxy.As<v8::Object>(), main_context))
+        !DeepFreeze(proxy.As<v8::Object>(), electron_isolated_context))
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:610]:
```suggestion
        !DeepFreeze(proxy.As<v8::Object>(), target_context))
```

File: spec/api-context-bridge-spec.ts [hunk 14]
```diff
@@ -114,6 +126,26 @@ describe('contextBridge', () => {
         expect(result).to.equal(123);
       });
 
+      it('should proxy numbers', async () => {
```
Comment (samuelmaddock) [inline spec/api-context-bridge-spec.ts:129]: Should be more descriptive. There's already a test using "should proxy numbers"

File: spec/api-context-bridge-spec.ts [hunk 15]
```diff
@@ -114,6 +126,26 @@ describe('contextBridge', () => {
...
+        await makeBindingWindow(() => {
+          contextBridge.exposeInIsolatedWorld(1004, 'example', 123);
+        }, 1004);
+        const result = await callWithBindings((root: any) => {
+          return root.example;
+        }, 1004);
+        expect(result).to.equal(123);
+      });
+
+      it('should proxy numbers', async () => {
```
Comment (samuelmaddock) [inline spec/api-context-bridge-spec.ts:139]: Duplicate test name

File: spec/api-context-bridge-spec.ts [hunk 16]
```diff
@@ -172,6 +204,18 @@ describe('contextBridge', () => {
         expect(result).to.equal('my-words');
       });
 
+      it('should proxy nested strings', async () => {
```
Comment (samuelmaddock) [inline spec/api-context-bridge-spec.ts:207]: Duplicate test name

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 17]
```diff
@@ -560,50 +560,57 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
     v8::MaybeLocal<v8::Value> maybe_proxy = PassValueToOtherContext(
-        isolated_context, main_context, api, &object_cache, false, 0);
+        electron_isolated_context, context, api, &object_cache, false, 0);
     if (maybe_proxy.IsEmpty())
       return;
     auto proxy = maybe_proxy.ToLocalChecked();
 
     if (base::FeatureList::IsEnabled(features::kContextBridgeMutability)) {
-      global.Set(key, proxy);
+      isolated.Set(key, proxy);
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:605]:
```suggestion
      properties.Set(key, proxy);
```

**Review by samuelmaddock — CHANGES_REQUESTED** [review]

File: docs/api/context-bridge.md [hunk 18]
```diff
@@ -46,6 +46,12 @@ The `contextBridge` module has the following methods:
 * `apiKey` string - The key to inject the API onto `window` with.  The API will be accessible on `window[apiKey]`.
 * `api` any - Your API, more information on what this API can be and how it works is available below.
 
+### `contextBridge.exposeInIsolatedWorld(worldId, apiKey, api)`
+
+* `worldId` Integer - The ID of the world to inject the API into. This has to be an existing world. `0` is the default world, `999` is the world used by Electron's `contextIsolation` feature. Using 999 would expose the object for preload context.
```
Comment (samuelmaddock) [inline docs/api/context-bridge.md:51]: Can we write a test to ensure no crash occurs when exposing an API in a world which doesn't exist yet?

Comment (samuelmaddock) [reply docs/api/context-bridge.md:51]: If it still works, maybe we can update these docs to remove the "existing world" part.

File: shell/renderer/api/electron_api_context_bridge.cc [hunk 19]
```diff
@@ -561,19 +561,25 @@ v8::MaybeLocal<v8::Object> CreateProxyForAPI(
...
   auto* render_frame = GetRenderFrame(isolate->GetCurrentContext()->Global());
   CHECK(render_frame);
   auto* frame = render_frame->GetWebFrame();
   CHECK(frame);
-  v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
-  gin_helper::Dictionary global(main_context->GetIsolate(),
-                                main_context->Global());
+
+  v8::Local<v8::Context> target_context =
+      world_id == 0 ? frame->MainWorldScriptContext()
```
Comment (samuelmaddock) [inline shell/renderer/api/electron_api_context_bridge.cc:578]: We have a constant for world IDs in https://github.com/electron/electron/blob/main/shell/common/world_ids.h which we could maybe make use of here.

**Review by samuelmaddock — COMMENTED** [review]
Code is looking good. Just had a small request for tests

**Review by itsananderson — APPROVED** [review]
API LGTM

File: shell/renderer/api/electron_api_context_bridge.cc (hunk 19 above)
Comment (RaisinTen) [reply shell/renderer/api/electron_api_context_bridge.cc:578]:
```suggestion
      world_id == WorldIDs::MAIN_WORLD_ID ? frame->MainWorldScriptContext()
```
and applying the cpplint diff should resolve this.

File: docs/api/context-bridge.md (hunk 18 above)
Comment (RaisinTen) [reply docs/api/context-bridge.md]: There's a test that uses world id `1004`, resolving.

**Review by RaisinTen — APPROVED** [review]
LGTM

**Review by samuelmaddock — APPROVED** [review]

**Comment by ckerr** [comment]
Not sure what's going on with the cation bot, but this has LGTM API signoff from 3 API WG members

**Comment by theogravity** [comment]
The documentation is lacking a proper example of how to use `exposeInIsolatedWorld()`. Do I attach the `worldId` to a `BrowserWindow` somewhere so that window has access to the isolated methods?

**Comment by qdirks** [comment]
@theogravity

> The documentation is lacking a proper example of how to use exposeInIsolatedWorld().

I struggled with this as well, but I finally found some context here [32588][1], which is the original feature request which this pull request closes.

> Do I attach the worldId to a BrowserWindow somewhere so that window has access to the isolated methods?

I had a similar misunderstanding. The `worldId` is attached to the `BrowserWindow` through the use of this method. Bear in mind that this is running in the preload script of the browser window, thus it is known that it will attach to that browser window. The real question is, "How do I run a script in the isolated world to access the exposed API?", right?

If so, then the method you're looking for is `webFrame.executeJavascriptInIsolatedWorld`, available from `const {webFrame} = require('electron');`, documentation [here][2]. To my understanding, `exposeInIsolatedWorld` can also be used to expose APIs to extensions.

If not, then what you're probably trying to do is expose the API to your scripts which run in the main world, e.g. you want to expose an API to the script tags in your html which you have loaded. If so, then use `exposeInMainWorld`.

Hope this helps.

[1]: https://github.com/electron/electron/issues/32588
[2]: https://www.electronjs.org/docs/latest/api/web-frame#webframeexecutejavascriptinisolatedworldworldid-scripts-usergesture-callback
